data/lazy_list/basic.lean

```diff
@@ -0,0 +1,29 @@
+/-
+Copyright (c) 2018 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Simon Hudon
+
+Traversable instance for lazy_lists.
+-/
+
+import data.traversable.equiv data.lazy_list
+
+namespace lazy_list
+
+open function
+
+def list_equiv_lazy_list (α : Type*) : list α ≃ lazy_list α :=
+by { refine { to_fun := lazy_list.of_list, inv_fun := lazy_list.to_list, .. };
+     simp [left_inverse,function.right_inverse];
+     intros x ,
+     all_goals
+     { induction x; [ refl, simp! [*] ] },
+     ext, rw [punit_eq_punit x], }
+
+instance : traversable lazy_list :=
+equiv.traversable list_equiv_lazy_list
```

Comment:
This is not a good implementation of traversable, because it only works on finite lazy_lists. You should write an implementation that works lazily. (Similar points apply for other uses of `equiv.traversable` - in most cases this produces unnecessarily inefficient code. Maybe you can shortcut the `is_lawful` part of the proof using the equiv, but the operations themselves should depend on the data structure.)

Reply:
`lazy_list` does not allow infinite lists. But it is true that this is stricter than necessary. As you're alluding to, I'm thinking of a way of using the `equiv` to prove the laws, provide a hand crafted `traverse` implementation with a proof of equivalence with `list.traverse` (in this case). I think we should merge this PR before I get started on that. I'm not sure yet of all the details.

Reply:
It `meta`-allows infinite lists, i.e. `lazy_list.iota`. I'm fine with postponing this to the next commit though.

Reply:
👍 Beside `reader`, I'm wondering what `monad` / `applicative` this will be usable with. Maybe `writer` with a lazy data structure?

Reply:
Note: If you want to (partly) traverse an infinite list (be it `lazy_list` or `stream`) in the `tactic` monad or other stateful monads, I have the `pipes` library that I'm working on and that handles traversal using co-routine-like machinery. 

Using `traversable` on `lazy_list` in the `tactic` monad will still result in trying to evaluate the whole list (even if we specialize `traverse`)